WordPress/gutenberg
# List View: Scroll selected block into view when single block selection changes

## What?


Fixes #46828, fixes #30432

While the list view is open, scroll the selected item into view if it is not already in view.

A caveat on the approach in this PR: I'm not sure how performant it is, so very happy for any feedback on this one!

## Why?


As described in #46828, with the list view open and a long post or page, it can be confusing if you change the selection within the editor canvas, but the list view does not scroll that block into view. It can be quite easy to lose your place on longer posts.

## How?


* Add a hook to the list view that scrolls the selected block's element into view if it isn't already visible.
* Use this at the `ListViewBlock` level, and also at the `ListViewBranch` level to factor in very long posts/pages where the placeholder for the selected block falls outside of the current window where "real" list view blocks are rendered. This way, if a user selects a block that is outside of the currently windowed area, we can still scroll it into view.

## Testing Instructions





In a really long post or page, with the list view opened, scroll the editor canvas to far down the page or post and select a block outside of the currently visible area within the list view. The list view should scroll the selected block into view.

To test the windowed area versus the non-windowed area, add a very large number of blocks (e.g. 60+) and try selecting from the very top block to the very bottom block in the list, via the editor canvas.

### Testing Instructions for Keyboard


It should be possible to test this via keyboard – with the list view opened and focus on the editor canvas, press the down arrow to navigate down blocks, and as you scroll past the bottom of the screen, or past the top of the screen, the list view should always display the selected block.

## Screenshots or screencast 

In the below screengrab, with the list view open, we go from the first block being selected, to scrolling the editor canvas and selecting a block that is outside of the visible area of the list view. When the block is selected, the list view is scrolled so that the selected block is at the top of the visible area of the list view.

https://user-images.githubusercontent.com/14988353/210718105-0f735197-31f7-461d-ad5c-114a36c88d94.mp4
